src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,733 @@
+                sz = (size_t)cast<ConstantInt>(call->getArgOperand(1))->getZExtValue();
+            }
+            else {
+                continue;
+            }
+            allocs[call] = sz;
+        }
+    }
+
+    auto &entry = F.getEntryBlock();
+    auto first = &entry.front();
+    CheckInstStack check_stack;
+    ReplaceUsesStack replace_stack;
+    std::set<Instruction*> alloc_uses;
+    LifetimeMarker lifetime(*this);
+    for (auto it: allocs) {
+        bool ignore_tag = true;
+        auto orig = it.first;
+        if (optimize && checkInst(orig, check_stack, alloc_uses, ignore_tag)) {
+            // The allocation does not escape or get used in a phi node so none of the derived
+            // SSA from it are live when we run the allocation again.
+            // It is now safe to promote the allocation to an entry block alloca.
+            size_t sz = it.second;
+            size_t align = 1;
+            // TODO make codegen handling of alignment consistent and pass that as a parameter
+            // to the allocation function directly.
+            if (!ignore_tag) {
+                align = sz <= 8 ? 8 : 16;
+                sz += align;
+            }
+            else if (sz >= 16) {
+                align = 16;
+            }
+            else if (sz >= 8) {
+                align = 8;
+            }
+            else if (sz >= 4) {
+                align = 4;
+            }
+            else if (sz >= 2) {
+                align = 2;
+            }
+            const auto &dbg = orig->getDebugLoc();
+#if JL_LLVM_VERSION >= 50000
+            Instruction *ptr = new AllocaInst(T_int8, 0, ConstantInt::get(T_int32, sz),
+                                              align, "", first);
+#else
+            Instruction *ptr = new AllocaInst(T_int8, ConstantInt::get(T_int32, sz),
+                                              align, "", first);
+#endif
+            lifetime.insert(ptr, ConstantInt::get(T_int64, sz), orig, alloc_uses);
+            ptr->setDebugLoc(dbg);
+            if (!ignore_tag) {
+                ptr = GetElementPtrInst::CreateInBounds(T_size, ptr,
+                                                        {ConstantInt::get(T_int32, align)}, "",
+                                                        first);
+                ptr->setDebugLoc(dbg);
+            }
+            auto cast = new BitCastInst(ptr, T_pjlvalue0, "", first);
+            cast->setDebugLoc(dbg);
+            // Someone might be reading the tag, initialize it.
+            if (!ignore_tag) {
+                auto tag = orig->getArgOperand(2);
+                auto cast2 = new BitCastInst(ptr, T_ppjlvalue0, "", orig);
+                cast2->setDebugLoc(dbg);
+                auto tagaddr = GetElementPtrInst::Create(T_pjlvalue, cast,
+                                                         {ConstantInt::get(T_size, -1)},
+                                                         "", orig);
```

Comment:
Yeah, It was tested but not after making it conditional....... I'm actually not sure how to actually trigger this....